starrSeq/binCells.go

```diff
@@ -0,0 +1,156 @@
+package starrSeq
+
+import (
+	"fmt"
+	"github.com/vertgenlab/gonomics/fileio"
+	"github.com/vertgenlab/gonomics/numbers"
+	"math/rand"
+	"sort"
+)
+
+// DetermineBin takes in an int and the probablity of any given bin and returns an int corresponding to the bin the cell belongs to.
+func DetermineBin(numBins int, prob float64) int {
+	var bin int
+
+	randNum := rand.Float64()      // draw a random number
+	for j := 0; j < numBins; j++ { //iterate over the number of bins
+		if j == numBins { //if we've gotten to the last bin and the cell still hasn't been partitioned, we don't have to check anything else we can just put it there.
+			return j
+		}
+		if randNum >= prob*float64(j) && randNum < prob*float64(j+1) { // check to see if our random number belongs to that bin
+			bin = j
+		}
+	}
+	return bin
+}
+
+// DistributeCells takes in a slice of string that is created in parseBam that has a list of cellBarcodes and constructs. It will partition those cells and constructs into separate slices. The number of bins is a user-input variable
+func DistributeCells(s ScStarrSeqSettings, cellTypeSlice []Read, fromDB bool) [][]Read {
+	var currCell string
+	var bin, count int
+	var found bool
+
+	whichBin := 'A'                   //for counting cells per bin
+	whichBinMap := make(map[rune]int) //for counting cells per bin
+
+	binnedCells := make([][]Read, s.BinCells) //make a slice of slice of string with the size of the user-specified number of bins
+	prob := 1.0 / float64(s.BinCells)         // determine the probability that the cell belongs to a particular bin
+
+	SortReadByCellBx(cellTypeSlice) //sort the slice of strings containing (cellBarcode \t construct) so that indentical cell barcodes line up next to one another
+	for _, i := range cellTypeSlice {
+		if i.Construct == "" {
+			fmt.Println("distribute cells empty")
+		}
+		if i.Bx == currCell { // if the cell barcode is the same as the one the loop just saw, partition that cell-construct pair into the same bin as the one before
+			binnedCells[bin] = append(binnedCells[bin], i)
+			continue
+		}
+		bin = DetermineBin(s.BinCells, prob)           //function to determine which bin to put the new cell into.
+		binnedCells[bin] = append(binnedCells[bin], i) //put cell into correct bin
+		currCell = i.Bx                                //set the cell barcode that was just partitioned to current cell for the next iteration
+		count, found = whichBinMap[whichBin+rune(bin)]
+		if !found {
+			whichBinMap[whichBin+rune(bin)] = 1
+		} else {
+			whichBinMap[whichBin+rune(bin)] = count + 1
+		}
+	}
+
+	var outSlice []string
+	if !fromDB {
+		for i := range whichBinMap {
+			outSlice = append(outSlice, fmt.Sprintf("Bin: %c\tcells: %d", i, whichBinMap[i]))
+		}
+		sort.Strings(outSlice)
+		for _, i := range outSlice {
+			fmt.Println(i)
+		}
+	}
+	return binnedCells
+}
+
+// DetermineIdealBins will determine how many bins should be used for the pseudobulk -binCells option.
```

Comment:
This is not a super well thought out idea, but since you're trying to get the ideal bin number mainly to make sure you have some reads from each negative control construct in each bin, could it make sense to just make this the bin number equal to the minimum number of cells for any of the constructs? For example, if you have NCa with 5 cells, NCb with 7 cells, and NCc with 8 cells, choose 5 bins to ensure you have at least one cell from each NC construct in each bin. Perhaps the main downside here is that you'd need to write a separate bin distribution function for the NC cells to make sure that each NC cell for a given construct will go to a new bin before you start repeating bins. I feel like that might end up being more efficient then simulating different bin sizes many times though.